### ISSUE TITLE
JSON-LD 

### ISSUE THREAD

**Issue by fimbault** [pchampin/sophia_rs#88]
Hi, 

It's great to get json-ld 1.1 support, at least for serialization and parsing. Even if it's early stage, it could be interesting to list it on https://json-ld.org
What's the long term objective in terms of compliance with the rest of the spec? (flattening, compaction, framing)

Fabien

**Comment by pchampin** [comment]
> What's the long term objective in terms of compliance with the rest of the spec? (flattening, compaction, framing)

Most algorithms in the spec take JSON as input and return JSON as output, so I consider them as slightly out-of-scope for Sophia (which is centered on the notion of RDF Graph). The only relevant algorithms for Sophia are therefore `toRdf` (i.e. a JSON-LD parser) and `fromRdf` (i.e. a JSON-LD serializer).

That being said, these two algorithms are not totally independant from the others:
* `toRdf` explicitly relies on the expansion algorithm
* `fromRdf` as described by the spec produces expanded JSON-LD, but a full-featured serializer should provide options to produce a compact JSON-LD document, according to a given context.

The way I see things:
* a separate crate will implement the JSON-LD algorithms according to the JSON-LD spec (I have [this one](https://github.com/timothee-haudebourg/json-ld) in mind, I am in contact with @timothee-haudebourg);
* `sophia_jsonld` will depend on this crate to provide implementations of the Sophia traits for parsers and serializer.

**Comment by roper79** [comment]
Hi, is there any progress/plans with a JSON-LD parser front? Thanks!

**Comment by pchampin** [comment]
at long (long) last, there is indeed some progress! d6d73d7

Long story short: there is now a `sophia_jsonld` crate providing a JSON-LD parser and a JSON-LD serializer (producing only the expanded form at the moment).

**Comment by damooo** [comment]
@pchampin Could you release the update to crates.io?

**Comment by pchampin** [comment]
> Could you release the update to crates.io?

I still have a few adjustments to make, but I will eventually.

Note that you can still use the commit as a dependency in cargo with
```
sophia = { git = "https://github.com/pchampin/sophia_rs.git", rev = "d6d73d78" }
```

Of course, you don't want to release on crates.io a crate with this kind of  dependencies (I guess the server won't let you, anyway), but for a work in progress, that's a good solution.

**Comment by pchampin** [comment]
I noticed that, and it should now be fixed.
The problem was :
- you need to enable the `jsonld` feature to have JSON-LD support
- the line causing the error with `futures_util` should have been included *only* when the feature is enabled, because `futures_util` is not included otherwise.

With the latest commit, the crate should compile with or withour the `jsonld` feature enabled. But of course, in your case, you should enable it!

Please close this issue if the problem is indeed solved, otherwise let me know and I'll try to figure it out.

**Comment by pchampin** [comment]
> @pchampin Could you release the update to crates.io?

@damooo by the way, I forgot to mention that the JSON-LD support was included in [version 0.8.0-alpha.3](https://crates.io/crates/sophia_api/0.8.0-alpha.3

Regarding the initial suggestion of this issue (adding Sophia to the list of JSON-LD implementations on json-ld.org): since Sophia is based on https://github.com/timothee-haudebourg/json-ld, and since that implementation is already mentioned there, I don't think it is appropriate to add Sophia to that list.

**Comment by damooo** [comment]
> JSON-LD support was included in [version 0.8.0-alpha.3]

Saw that already _/\\\_.

Now basic support is added in [manas](https://github.com/manomayam/manas/pull/35).

It also adds json-ld support to `rdf_dynsyn` crate (not yet updated in crates.io), to be dynamically specified. And also includes a custom http loader, that allows for caching/any-middleware, and fixes [few issues](https://github.com/timothee-haudebourg/json-ld/issues/69) in existing loader.

**Comment by LuisOsta** [comment]
Thanks!
